### DISCUSSION TITLE
Will there be something new to replace XAML?

### DISCUSSION THREAD

**Discussion by EarzuChan** [dotnet/maui#23757] [Ideas]
### Description

XAML and this old form of UI description and binding might have been novel in 2006, but by 2024, it might need to change. Has your team considered learning from Jetpack Compose, Vue.js, and React to provide a way to declare UI in a reactive, data-driven, and function-compositional manner?

### Public API Changes

Let me show you a comparison:

# XAML Example

```xml
<Window x:Class="ExampleApp.MainWindow"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        Title="MainWindow" Height="350" Width="525">
    <Grid>
        <TextBlock Text="{Binding Greeting}" />
        <Button Content="Click Me" Command="{Binding ClickCommand}" />
        <TextBlock Text="Visibility Controlled Text" Visibility="{Binding IsVisible, Converter={StaticResource BooleanToVisibilityConverter}}" />
    </Grid>
</Window>
```

# Jetpack Compose Example

```kotlin
@Composable
fun MainWindow(viewModel: MainViewModel) {
    val greeting by viewModel.greeting.collectAsState()
    val isVisible by viewModel.isVisible.collectAsState()

    Column {
        Text(text = greeting)
        Button(onClick = { viewModel.onClick() }) {
            Text("Click Me")
        }
        if (isVisible) {
            Text(text = "Visibility Controlled Text")
        }
    }
}
```

### Intended Use-Case

This is a very modern, progressive, simple (Vue has its own SFC mode, while Compose eliminates the gap between the code layer and the view layer), and efficient idea. Has your team considered this? Or is your team already working on it or has a similar implementation? Please let me know.

**Comment by EarzuChan** [comment]
Additionally, I'm curious to hear your thoughts, but from a code perspective, Vue's computed properties and watchEffect are truly amazing. They make state-driven UI rendering seamless and handle data effortlessly.
This means I don't have to worry about updating the UI manually, as my data updates automatically and naturally.
Moreover, the template syntax makes data binding and event handling incredibly convenient and beautiful.
Although I am a staunch .NET supporter, for my latest project, I chose to write the backend logic in C# and use a Vue 3 + TypeScript interface displayed in WPF + WebView2 for the frontend.

**Comment by EarzuChan** [comment]
Okay, I found dotnet/Comet, but it seems dead

**Comment by giuseppenovielli** [comment]
https://github.com/adospace/reactorui-maui

**Comment by akhanalcs** [comment]
I also want a replacement for XAML but the new format can be something like Blazor-syntax or just pure C# way of writing UI.

Take a look at these:

https://github.com/microsoft/microsoft-ui-xaml/issues/2499 (I like this a lot)

https://github.com/dotnet/aspnetcore/issues/44866

**Reply by EarzuChan** [reply]
True dude! Take the advantages from Compose and Vue.
